In [101]:
import pandas as pd 
import numpy as np

Criando a função para relatório inicial das bases

In [102]:
def relatorio_inicial(df):
    """
    PRE AVALIACAO DA BASE
    """
    print("=" * 48)
    print(f"ETAPA 1 - AUDITORIA")
    print("=" * 48)
    
    print("\n[1] COLUNAS DA BASE:")
    print(df.columns.tolist())
    
    print("\n[2] INFO DA BASE:")
    df.info()
    
    print("\n[3] TIPOS DE DADOS (dtypes):")
    print(df.dtypes)
    
    print("\n[4] ESTATÍSTICAS DESCRITIVAS (Numéricas):")
    print(df.describe())

    '''print("\n[5] ESTATÍSTICAS DESCRITIVAS (Objetos/Texto): ")
    try:
        print(df.describe(include='object'))
    except ValueError:
        print("Não há colunas de texto nesta base.")'''
    
    print("\n[6] QUANTIDADE DE NULOS POR COLUNA:")
    print(df.isnull().sum())
     
    print("\n[8] PERCENTUAL DE NULOS (> 15%):")
    percentual_nulos = (df.isnull().sum() / len(df) * 100).round(1)
    print(percentual_nulos)
        
    print("\n[9] LINHAS COM VALORES NULOS:")
    qtd_nulos_linhas = df.isnull().any(axis=1).sum()
    print(f"Total de linhas com pelo menos um valor nulo: {qtd_nulos_linhas}")
    
    print("\n[10] TOTAL DE LINHAS DUPLICADAS:")
    total_duplicadas = df.duplicated().sum()
    print(f"Total: {total_duplicadas}")
    
    print("\n" + "=" * 48)
    print(f"FIM DO RELATÓRIO")
    print("=" * 48 + "\n")

def relatorio_final(df_original, df_atual):
    """
    RELATÓRIO FINAL DE AUDITORIA
    """
    registros_iniciais = df_original.shape[0]
    registros_finais = df_atual.shape[0]
    colunas_iniciais = df_original.shape[1]
    colunas_finais = df_atual.shape[1]

    print('=' * 48)
    print('      RELATÓRIO FINAL DE AUDITORIA')
    print('=' * 48)
    print(f'Registros iniciais:     {registros_iniciais:>6}')
    print(f'Registros finais:       {registros_finais:>6}')
    print(f'Registros removidos:    {registros_iniciais - registros_finais:>6}')
    print(f'Colunas originais:      {colunas_iniciais:>6}')
    print(f'Colunas finais:         {colunas_finais:>6}')
    print('=' * 48)
    print(f'Nulos restantes:        {df_atual.isnull().sum().sum():>6}')
    print(f'Duplicatas restantes:   {df_atual.duplicated().sum():>6}')
    print('=' * 48)

## Convertendo as bases em Dataframe

Concatenando base adicional de clientes à base original 2023-2024

In [103]:
canais = pd.read_csv('canais_venda.csv', sep=';', encoding='utf-8')
clientes = pd.read_csv('clientes.csv', sep=';', encoding='utf-8')
funcionarios = pd.read_csv('funcionarios.csv', sep=';', encoding='utf-8')
reservas = pd.read_csv('reservas.csv', sep=';', encoding='utf-8')
quartos = pd.read_csv('tipos_quarto.csv', sep=';', encoding='utf-8')
unidades = pd.read_csv('unidades.csv', sep=';', encoding='utf-8')

#Adicionando clientes novos às bases
clientes_novos=pd.read_csv('clientes_novos.csv', sep=';', encoding='utf-8')
df_concatenado_clientes = pd.concat([clientes, clientes_novos], ignore_index=True)

quartos_novos=pd.read_csv('tipos_quarto_novos.csv', sep=';', encoding='utf-8')
df_concatenado_quartos = pd.concat([quartos, quartos_novos], ignore_index=True)

funcionarios_novos=pd.read_csv('funcionarios_novos.csv', sep=';', encoding='utf-8')
df_concatenado_funcionarios = pd.concat([funcionarios, funcionarios_novos], ignore_index=True)

reservas_novos=pd.read_csv('reservas_novas.csv', sep=';', encoding='utf-8')
df_concatenado_reservas = pd.concat([reservas, reservas_novos], ignore_index=True)

unidades_total=pd.read_csv('unidades_novos.csv', sep=';', encoding='utf-8')
df_concatenado_unidades = pd.concat([unidades, unidades_total], ignore_index=True)

In [104]:
# df original - criando as cópias para o relatório final

canais_original = canais.copy()

df_concatenado_unidades_original = df_concatenado_unidades.copy()
df_concatenado_clientes_original = df_concatenado_clientes.copy()
df_concatenado_funcionarios_original = df_concatenado_funcionarios.copy()
df_concatenado_quartos_original = df_concatenado_quartos.copy()
df_concatenado_reservas_original = df_concatenado_reservas.copy()

## Etapa 1 - Auditoria

Aplica a função relatório inicial. Identifica inconsistências iniciais na base

In [105]:
#relatorio_inicial (canais)
#relatorio_inicial (funcionarios)
#relatorio_inicial(reservas)
#relatorio_inicial(quartos)
#relatorio_inicial(unidades)
#relatorio_inicial(clientes_novos)

## 2. Limpeza e tratamento

### 2.1 Base: funcionários

Atualização e Correção da Base Salarial com map

In [106]:
novos_salarios = {
    9: '1870.00',
    10:'2314.00',
    14: '5480.00',
    37: '2950.00',
    43:'5600.00',
    92:'8600.00',
    104:'2890.00',
    11:'2800.00',
    26:'2150.00',
}

df_concatenado_funcionarios.loc[df_concatenado_funcionarios['id_funcionario'].isin(novos_salarios), 'salario'] = df_concatenado_funcionarios['id_funcionario'].map(novos_salarios)

In [107]:
df_concatenado_funcionarios['departamento'].value_counts()

departamento
Governança             33
Manutenção             24
Segurança              20
Financeiro             15
Recepção               15
Alimentos e Bebidas    13
Gov.                    8
Administração           8
A&B                     5
Admin.                  1
Operações               1
operações               1
Name: count, dtype: int64

In [108]:
df_concatenado_funcionarios['departamento'].value_counts()

mapa = {
    'Gov.':                     'Governança',
    'Admin.':  'Administração'             ,
    'operações' :'Operações',
    'A&B': 'Alimentos e Bebidas'

}

df_concatenado_funcionarios['departamento'] = df_concatenado_funcionarios['departamento'].map(mapa).fillna(df_concatenado_funcionarios['departamento'])#joga as informações do dicionario mapa no df. filne mantem os valores corretos que não precisam de substituição

print('Após mapeamento:\n')
print(sorted(df_concatenado_funcionarios['departamento'].unique()))

Após mapeamento:

['Administração', 'Alimentos e Bebidas', 'Financeiro', 'Governança', 'Manutenção', 'Operações', 'Recepção', 'Segurança']


Ajustando e padronizando strings nas colunas

In [109]:
df_concatenado_funcionarios['salario'] = df_concatenado_funcionarios['salario'].str.replace('"', '', regex=False).str.replace("'","", regex=False).str.replace(",",".", regex=False)

- Convergindo colunas com datas para Datetime
- Aplicando função para correção da coluna salário
- 

In [110]:
df_concatenado_funcionarios['data_admissao'] = pd.to_datetime(
    df_concatenado_funcionarios['data_admissao'],
    format='mixed',
    dayfirst=True,
    errors='coerce'
)

def corrigir_valor(x):
    if pd.isna(x):
        return x
    x = str(x)
    if x.count('.') == 2:
        x = x.replace('.', '', 1)
    return x

df_concatenado_funcionarios['salario'] = df_concatenado_funcionarios['salario'].apply(corrigir_valor)

df_concatenado_funcionarios['salario'] = pd.to_numeric(df_concatenado_funcionarios['salario'], errors='coerce')

df_concatenado_funcionarios[df_concatenado_funcionarios['salario'].isna()][['nome', 'salario']]

novos_salarios = {
    142: 7850.00    
}

df_concatenado_funcionarios.loc[df_concatenado_funcionarios['id_funcionario'].isin(novos_salarios), 'salario'] = df_concatenado_funcionarios['id_funcionario'].map(novos_salarios)

df_concatenado_funcionarios['cargo'] = df_concatenado_funcionarios['cargo'].str.upper()

df_concatenado_funcionarios.head(50)  

,id_funcionario,id_unidade,nome,cargo,departamento,salario,data_admissao
0,1,1,Gabriela Lima,AUXILIAR DE LIMPEZA,Governança,2139.0,2020-05-10
1,2,1,Abel Correia,SEGURANÇA,Segurança,1746.0,2021-12-29
2,3,1,Sara Teixeira,ANALISTA FINANCEIRO,Financeiro,5356.0,2022-01-03
3,4,1,Julia Monteiro,AUXILIAR DE LIMPEZA,Governança,2070.0,2019-10-06
4,5,1,Daniela Ramos,AUXILIAR DE LIMPEZA,Governança,1572.0,2018-10-16
5,6,1,Lucas Souza,AUXILIAR DE LIMPEZA,Governança,1955.0,2024-01-02
6,7,1,Wilson Duarte,RECEPCIONISTA,Recepção,1906.0,2023-03-24
7,8,1,Nicolas Cunha,RECEPCIONISTA,Recepção,2417.0,2019-02-07
8,9,1,Bruno Costa,AUXILIAR DE LIMPEZA,Governança,1870.0,2021-02-11
9,10,1,Bruno Costa,MANUTENÇÃO,Manutenção,2314.0,2022-11-28


### 2.2 base: Reservas

Execuções principais

1) correção das diárias
2) correção da avaliação dos hóspedes

* Considerando a necessidade de métricas posteriores e decisão do cliente em
manter nulos valores iguais a 0, optou-se por manter valores nulos/impossíveis
como não avaliados. Contudo, vale alertar novamente que os valores não informados somam-se à 
 646 pontos elevando-se à 13.9% do total de avaliações.


*Inclusão e correção de valores fornecidos pelo cliente com função MAP.

*Correção do sinal negativo na quantidade de hóspedes. 

In [111]:
relatorio_inicial(df_concatenado_reservas)

ETAPA 1 - AUDITORIA

[1] COLUNAS DA BASE:
['id_reserva', 'id_unidade', 'id_tipo_quarto', 'id_cliente', 'id_canal', 'data_checkin', 'data_checkout', 'qtd_diarias', 'num_hospedes', 'avaliacao_hospede', 'status_reserva', 'forma_pagamento']

[2] INFO DA BASE:
<class 'pandas.DataFrame'>
RangeIndex: 4664 entries, 0 to 4663
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id_reserva         4664 non-null   int64  
 1   id_unidade         4664 non-null   int64  
 2   id_tipo_quarto     4664 non-null   int64  
 3   id_cliente         4664 non-null   int64  
 4   id_canal           4455 non-null   float64
 5   data_checkin       4664 non-null   str    
 6   data_checkout      4664 non-null   str    
 7   qtd_diarias        4664 non-null   int64  
 8   num_hospedes       4664 non-null   int64  
 9   avaliacao_hospede  4018 non-null   float64
 10  status_reserva     4664 non-null   str    
 11  forma_pagamento    

In [112]:
novos_valores = {
    728: 7, 955: 4, 1134: 3, 1152: 2,
    1448: 6, 1581: 4, 2312: 4, 2384: 2,
    2509: 1, 4030: 1, 3472: 1, 3505: 1, 
    3838: 1, 4048: 1
}
novas_avaliacoes = {
    37: 1, 902: 1, 2073: 1, 933: 3, 1718: 3, 2167: 3,
    3541: 1, 2874: 1, 4127: 1, 4088: 3
}

correcao_novos_hospedes = {
    568: 1, 675: 2, 903: 1, 1201: 2, 1883: 2,
    2427: 2, 4463: 1, 4506: 2, 4040: 2, 2687: 2,
}

ids_hospedes = list(correcao_novos_hospedes.keys())
df_concatenado_reservas.loc[
    df_concatenado_reservas['id_reserva'].isin(ids_hospedes), 'num_hospedes'
] = df_concatenado_reservas['id_reserva'].map(correcao_novos_hospedes)

ids_avaliacoes = list(novas_avaliacoes.keys())

df_concatenado_reservas.loc[
    df_concatenado_reservas['id_reserva']
    .isin(ids_avaliacoes),
    'avaliacao_hospede'
] = df_concatenado_reservas['id_reserva'].map(novas_avaliacoes)

ids_valores = list(novos_valores.keys())

df_concatenado_reservas.loc[
    df_concatenado_reservas['id_reserva']
    .isin(ids_valores),
    'qtd_diarias'] = df_concatenado_reservas['id_reserva'].map(novos_valores)

df_filtrado = df_concatenado_reservas[df_concatenado_reservas['id_reserva'].isin(ids_valores)]

print((df_concatenado_reservas['num_hospedes'] < 0).sum())
print((df_concatenado_reservas['qtd_diarias'] < 0).sum())
print((df_concatenado_reservas['id_canal'] == -1).sum())

0
0
0


In [113]:
df_concatenado_reservas[df_concatenado_reservas['avaliacao_hospede'].isnull()]

,id_reserva,id_unidade,id_tipo_quarto,id_cliente,id_canal,data_checkin,data_checkout,qtd_diarias,num_hospedes,avaliacao_hospede,status_reserva,forma_pagamento
7,8,4,4,33,3.0,2023-08-26,2023-09-02,7,2,NaN,no-show,Dinheiro
13,14,4,3,68,3.0,2024-03-07,2024-03-11,4,3,NaN,cancelada,Dinheiro
14,15,6,2,173,1.0,2024-07-20,2024-07-23,3,1,NaN,cancelada,Transferência
35,36,11,1,259,4.0,2023-03-26,2023-03-29,3,1,NaN,cancelada,Cartão de Crédito
40,41,9,3,74,1.0,2023-12-03,2023-12-09,6,3,NaN,confirmada,Cartão de Débito
...,...,...,...,...,...,...,...,...,...,...,...,...
4646,3928,11,1,175,3.0,2025-03-20,2025-03-21,1,2,NaN,confirmada,Transferência
4648,3844,11,3,33,4.0,2026-01-26,2026-01-28,2,3,NaN,confirmada,Transferência
4649,2835,4,2,36,3.0,2025-07-05,2025-07-07,2,1,NaN,confirmada,Dinheiro
4651,4172,3,1,102,2.0,2025-04-01,2025-04-02,1,4,NaN,CONFIRMADA,Dinheiro


### Identifica as duplicidades

Refinamento e inspeção visual com a função *value.counts()*

In [114]:
#print('--- ativo ---')
#print(df_concatenado_reservas['id_reserva'].value_counts())

#print('--- ativo ---')
#print(df_concatenado_reservas['id_unidade'].value_counts())

#print('--- ativo ---')
#print(df_concatenado_reservas['id_tipo_quarto'].value_counts())

#print('--- ativo ---')
#print(df_concatenado_reservas['id_cliente'].value_counts())

#print('--- ativo ---')
#print(df_concatenado_reservas['status_reserva'].value_counts())

#print('--- ativo ---')
#print(df_concatenado_reservas['forma_pagamento'].value_counts())


* Remove as duplicidades na coluna Id reserva

* Resolve a duplicidades dos ids 2486, 2747, 2851, 3020, 3389, 3743, 3936, 4069, 4172, 4282, 4311, 4409, 4463, 4558, 4560 através da função Keep com a condição First que elimina
a segunda ocorrência.

In [115]:
shape_inicial = df_concatenado_reservas.shape[0]

df_concatenado_reservas = df_concatenado_reservas.drop_duplicates(
    subset=['id_reserva'], 
    keep='first'
    )

shape_final = df_concatenado_reservas.shape[0]

print(df_concatenado_reservas.shape)
print(f'Registros iniciais: {shape_inicial}')
print(f'Registros finais: {shape_final}')
print(f'{shape_inicial - shape_final} registros apagados')

(4649, 12)
Registros iniciais: 4664
Registros finais: 4649
15 registros apagados


2) Padronizando strings nas colunas **status da reserva** e **forma de pagamento**

In [116]:
#print('\n--- nomes com espaço nas bordas ---')

#mascara = df_concatenado_reservas['status_reserva'].str.startswith(' ') | df_concatenado_reservas['status_reserva'].str.endswith(' ')
#mascara = df_concatenado_reservas['forma_pagamento'].str.startswith(' ') | df_concatenado_reservas['forma_pagamento'].str.endswith(' ')

#df_concatenado_reservas['forma_pagamento'].value_counts()

#print(df_concatenado_reservas.loc[mascara, 'status_reserva'].tolist())
#print(df_concatenado_reservas.loc[mascara, 'forma_pagamento'].tolist())

In [117]:
df_concatenado_reservas['status_reserva'] = df_concatenado_reservas['status_reserva'].str.strip().str.upper()

df_concatenado_reservas['forma_pagamento'] = df_concatenado_reservas['forma_pagamento'].str.strip().str.upper()

df_concatenado_reservas[['forma_pagamento', 'status_reserva']].head(10)
#df_concatenado_reservas[['forma_pagamento', 'status_reserva']].tail(10)


,forma_pagamento,status_reserva
0,CARTÃO DE CRÉDITO,CONFIRMADA
1,PIX,CONCLUÍDA
2,CARTÃO DE CRÉDITO,CONCLUÍDA
3,PIX,CANCELADA
4,TRANSFERÊNCIA,CONFIRMADA
5,TRANSFERÊNCIA,CONFIRMADA
6,DINHEIRO,CONFIRMADA
7,DINHEIRO,NO-SHOW
8,DINHEIRO,CONCLUÍDA
9,PIX,CONFIRMADA


Padronização de strings na coluna STATUS utilizando a função Map

In [118]:
mapa_status = {
    'CONFIRMADA': 'CONFIRMADA',
    'CONF.': 'CONFIRMADA',
    'CANCELADA': 'CANCELADA',
    'CANCEL.': 'CANCELADA',
    'CONCLUÃDA': 'CONCLUÍDA',
    'CONCLUIDA': 'CONCLUÍDA',
    'NO-SHOW': 'NO-SHOW'
}

df_concatenado_reservas['status_reserva'] = df_concatenado_reservas['status_reserva'].map(mapa_status)

print(df_concatenado_reservas['status_reserva'].value_counts())

status_reserva
CONFIRMADA    3090
CANCELADA      652
NO-SHOW        242
CONCLUÍDA       32
Name: count, dtype: int64


Padronização de strings na coluna FORMA DE PAGAMENTO utilizando a função Map.

In [119]:
mapa_pagamento = {
    'DINHEIRO': 'DINHEIRO',
    'CASH': 'DINHEIRO',

    'PIX': 'PIX',

    'CARTÃO DE DÉBITO': 'CARTÃO DE DÉBITO',
    'CARTAO DEBITO': 'CARTÃO DE DÉBITO',
    'CD': 'CARTÃO DE DÉBITO',
    'DÉBITO': 'CARTÃO DE DÉBITO',

    'CARTÃO DE CRÉDITO': 'CARTÃO DE CRÉDITO',
    'CARTAO CREDITO': 'CARTÃO DE CRÉDITO',
    'C. CRÉDITO': 'CARTÃO DE CRÉDITO',
    'CRED': 'CARTÃO DE CRÉDITO',
    'CC': 'CARTÃO DE CRÉDITO',

    'TRANSFERÊNCIA': 'TRANSFERÊNCIA',
    'TRANSFERENCIA': 'TRANSFERÊNCIA',
    'TED': 'TRANSFERÊNCIA'
}

df_concatenado_reservas['forma_pagamento'] = df_concatenado_reservas['forma_pagamento'].map(mapa_pagamento)

print(df_concatenado_reservas['forma_pagamento'].value_counts())

forma_pagamento
DINHEIRO             978
CARTÃO DE DÉBITO     922
PIX                  920
CARTÃO DE CRÉDITO    908
TRANSFERÊNCIA        903
Name: count, dtype: int64


### Tratando nulos: recurso *fillna()* para o preenchimento de valores não informados

* Decisão de não exclusão devido ao tipo de dado (cadastro de clientes e reservas)

In [139]:

df_concatenado_reservas['id_canal'] = (
    df_concatenado_reservas['id_canal']
    .fillna(-1)
)
df_concatenado_reservas['forma_pagamento'] = (
    df_concatenado_reservas['forma_pagamento']
    .fillna('NÃO INFORMADO')
)
df_concatenado_reservas['status_reserva'] = (
    df_concatenado_reservas['status_reserva']
    .fillna('NÃO INFORMADO')
)

print(df_concatenado_reservas.isnull().sum())
#print(df_concatenado_reservas['id_canal'].dtype)
#print(df_concatenado_reservas['id_canal'].unique())

id_reserva             0
id_unidade             0
id_tipo_quarto         0
id_cliente             0
id_canal               0
data_checkin           0
data_checkout          0
qtd_diarias            0
num_hospedes           0
avaliacao_hospede    644
status_reserva         0
forma_pagamento        0
dtype: int64


Corrige a coluna de avaliação do hospede

* Decisão: o recurso fillna impede que o pandas calcule métrica posteriormente
* Procedimento: pd.to_numeric para alternar a coluna de avaliação ao formato numérico
* Resultado esperado: Pandas ignore valores nulos e permita análise métrica de avaliações.
Coluna avaliação retorna a float.


In [121]:
df_concatenado_reservas['avaliacao_hospede'] = pd.to_numeric(
    df_concatenado_reservas['avaliacao_hospede'], 
    errors='coerce')

print(df_concatenado_reservas['avaliacao_hospede'].dtype)
print(df_concatenado_reservas['avaliacao_hospede'].isnull().sum())
print(df_concatenado_reservas['avaliacao_hospede'].mean().round(2))

float64
644
6.68


### Corrigindo dtype das colunas
1) Corrige a coluna id_canal em formato INT

2) Coluna de datas de *STR* para *Datetime*


In [122]:
df_concatenado_reservas['id_canal'] = df_concatenado_reservas['id_canal'].astype(int)

In [123]:
df_concatenado_reservas['data_checkin'] = pd.to_datetime(
    df_concatenado_reservas['data_checkin'],
    format='mixed',
    dayfirst=True,
    errors='coerce'
)
df_concatenado_reservas['data_checkout'] = pd.to_datetime(
    df_concatenado_reservas['data_checkout'],
    format='mixed',
    dayfirst=True,
    errors='coerce'
)

print(df_concatenado_reservas[['data_checkin', 'data_checkout']].dtypes)
print("\n")
print(df_concatenado_reservas[['data_checkin', 'data_checkout']].isnull().sum())

data_checkin     datetime64[us]
data_checkout    datetime64[us]
dtype: object


data_checkin     0
data_checkout    0
dtype: int64


3) Tratamento dos valores negativos nas colunas: *id_canal, qtd_diarias e num_hospedes*

3.1 Aplicando filtro simples para identificar valores negativos

In [124]:
negativos_diarias = df_concatenado_reservas[df_concatenado_reservas['qtd_diarias'] < 0]
negativos_hospedes = df_concatenado_reservas[df_concatenado_reservas['num_hospedes'] < 0]

print(negativos_diarias)
print(negativos_hospedes)

Empty DataFrame
Columns: [id_reserva, id_unidade, id_tipo_quarto, id_cliente, id_canal, data_checkin, data_checkout, qtd_diarias, num_hospedes, avaliacao_hospede, status_reserva, forma_pagamento]
Index: []
Empty DataFrame
Columns: [id_reserva, id_unidade, id_tipo_quarto, id_cliente, id_canal, data_checkin, data_checkout, qtd_diarias, num_hospedes, avaliacao_hospede, status_reserva, forma_pagamento]
Index: []


Nesta etapa, aplicamos os filtros finais de consistência para garantir que nenhuma anomalia numérica (valores negativos, zeros ou exagerados) permaneça na base, validando as regras de negócio do hotel.

**O que este bloco faz:**
1. **Auditoria de Exclusão:** Salva o volume inicial de linhas para rastrear o impacto do filtro.
2. **Filtros de Regra de Negócio:**
   - `num_hospedes`: Mantém apenas reservas entre **1 e 20 hóspedes**.
   - `qtd_diarias`: Mantém apenas estadias entre **1 e 60 dias**.
3. **Contagem de Nulos Intencionais:** Verifica quantos registros ficaram sem `avaliacao_hospede` (mantidos como `NaN` propositalmente para não distorcer o cálculo de médias futuras).

In [125]:
antes = df_concatenado_reservas.shape[0]

df_concatenado_reservas = df_concatenado_reservas[
    df_concatenado_reservas['num_hospedes'].between(1, 20) &
    df_concatenado_reservas['qtd_diarias'].between(1, 60)
]

total_nulos = df_concatenado_reservas['avaliacao_hospede'].isnull().sum()

print(f'Valores removidos (fora do intervalo): {antes - df_concatenado_reservas.shape[0]}')
print(f'Shape atual: {df_concatenado_reservas.shape}')
print(f"Total de avaliações nulas: {total_nulos}")

Valores removidos (fora do intervalo): 0
Shape atual: (4649, 12)
Total de avaliações nulas: 644


Aplica as funções de relatórios. Cria o comparativo de execuções

In [126]:
df_atual = df_concatenado_reservas
df_original = df_concatenado_reservas_original

relatorio_final(df_original, df_atual)

      RELATÓRIO FINAL DE AUDITORIA
Registros iniciais:       4664
Registros finais:         4649
Registros removidos:        15
Colunas originais:          12
Colunas finais:             12
Nulos restantes:           644
Duplicatas restantes:        0


### 2.3 base: Clientes

1) Relatório Inicial

- Coluna Faixa Etária em formato STR
- Presença de valores nulos na mesma coluna
- Erros de grafia nas colunas 'Estado Origem', 'Pagamento' e Tipo Cliente
- Duplicidade do ID Cliente n° 47

Por decisão do cliente os registros (total 29) vazios nas colunas 'faixa_etária'
são mantidos como nulos (Nan) e não 0.

In [127]:
df_concatenado_clientes_original = df_concatenado_clientes.copy()

relatorio_inicial(df_concatenado_clientes_original)

ETAPA 1 - AUDITORIA

[1] COLUNAS DA BASE:
['id_cliente', 'nome', 'cidade_origem', 'estado_origem', 'faixa_etaria', 'tipo_cliente']

[2] INFO DA BASE:
<class 'pandas.DataFrame'>
RangeIndex: 321 entries, 0 to 320
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   id_cliente     321 non-null    int64
 1   nome           321 non-null    str  
 2   cidade_origem  321 non-null    str  
 3   estado_origem  321 non-null    str  
 4   faixa_etaria   290 non-null    str  
 5   tipo_cliente   321 non-null    str  
dtypes: int64(1), str(5)
memory usage: 15.2 KB

[3] TIPOS DE DADOS (dtypes):
id_cliente       int64
nome               str
cidade_origem      str
estado_origem      str
faixa_etaria       str
tipo_cliente       str
dtype: object

[4] ESTATÍSTICAS DESCRITIVAS (Numéricas):
       id_cliente
count  321.000000
mean   160.146417
std     92.592557
min      1.000000
25%     80.000000
50%    160.000000
75%    240.000000
max  

Refinamento da auditoria com a função *value.counts()* e filtro negativo

* Inconsistências de grafias stings
* Relação inconsistênte entre Estado de origem e cidade de origem divergente

In [128]:

#display(df_concatenado_clientes[df_concatenado_clientes['id_cliente'] < 0])

print("\n--- Estado de Origem ---")
#display(df_concatenado_clientes['estado_origem'].value_counts())

print("\n--- Faixa Etária ---")
#display(df_concatenado_clientes['faixa_etaria'].value_counts())

print("\n--- Tipo de Cliente ---")
#display(df_concatenado_clientes['tipo_cliente'].value_counts())

print("\n--- Cidade de Origem ---")
#display(df_concatenado_clientes['cidade_origem'].value_counts())

#display(df_concatenado_clientes['id_cliente'].value_counts())
#display(df_concatenado_clientes['nome'].value_counts())


--- Estado de Origem ---

--- Faixa Etária ---

--- Tipo de Cliente ---

--- Cidade de Origem ---


* Coluna Nome: Padronização dos strings

In [129]:
df_concatenado_clientes['nome'] = df_concatenado_clientes['nome'].str.strip().str.title()

df_concatenado_clientes['nome']

0             Otavio Reis
1           Larissa Sousa
2          Tatiana Coelho
3           Wilson Duarte
4           Gabriela Lima
              ...        
316    Priscila Guimarães
317       Rodrigo Vasques
318        Sabrina Leitão
319        Thiago Bicalho
320        Vivian Lacerda
Name: nome, Length: 321, dtype: str

* Coluna 'estado_origem': padronização dos strings e correspodência cidade Estado.

In [130]:
mapa_rio_de_janeiro = {'RIO DE JANEIRO': 'RJ', 'R.J.': 'RJ', 'Rio de Janeiro': 'RJ'}

mapa_cidade_estado = {
    'Fortaleza': 'CE',
    'São Paulo': 'SP',
    'Recife': 'PE',
    'Curitiba': 'PR',
    'Brasília': 'DF',
    'Belo Horizonte': 'MG',
    'Salvador': 'BA',
    'Rio De Janeiro': 'RJ',
    'Niterói': 'RJ',
    'Porto Alegre': 'RS',
    'Lisboa': 'EX',
    'Buenos Aires': 'EX',
    'Nova York': 'EX',
    'Miami': 'EX',
}

df_concatenado_clientes['estado_origem'] = df_concatenado_clientes['cidade_origem'].map(mapa_cidade_estado)

df_concatenado_clientes['estado_origem'] = df_concatenado_clientes['estado_origem'].replace(mapa_rio_de_janeiro)

print(df_concatenado_clientes[['cidade_origem', 'estado_origem']].isnull().sum())


cidade_origem     0
estado_origem    34
dtype: int64


Coluna 'Tipo_cliente' - padronização dos strings

In [131]:
df_concatenado_clientes['tipo_cliente'] = df_concatenado_clientes['tipo_cliente'].str.strip().str.upper()

mapa_tipo_cliente = {
    'CORPORATIVO': 'CORPORATIVO',
    'PF': 'PESSOA FÍSICA',
    'PJ': 'PESSOA JURÍDICA',
    'PESSOA FISICA': 'PESSOA FÍSICA',
    'PESSOA FÍSICA': 'PESSOA FÍSICA',
    'PESSOA JURÍDICA': 'PESSOA JURÍDICA',
}

df_concatenado_clientes['tipo_cliente'] = df_concatenado_clientes['tipo_cliente'].map(mapa_tipo_cliente)

print(df_concatenado_clientes['tipo_cliente'].value_counts())

tipo_cliente
PESSOA FÍSICA      150
CORPORATIVO        146
PESSOA JURÍDICA     25
Name: count, dtype: int64


Coluna 'cidade_origem': padronização dos strings
* coloca primeira letra maiuscula
* remove espaços antes/após strings

In [132]:
df_concatenado_clientes['cidade_origem'] = (
    df_concatenado_clientes['cidade_origem']
    .str.strip()
    .str.title()
)

df_concatenado_clientes['cidade_origem']

0           São Paulo
1            Curitiba
2      Rio De Janeiro
3            Brasília
4      Rio De Janeiro
            ...      
316            Lisboa
317    Rio De Janeiro
318          Brasília
319    Belo Horizonte
320      Buenos Aires
Name: cidade_origem, Length: 321, dtype: str

* Resolvendo a duplicidade de ID - n° 47

* Com base na recomendação do cliente, atribuiumos o ID 401 ao cliente Carlos Neto.

In [133]:
df_concatenado_clientes['id_cliente'].value_counts()

df_concatenado_clientes.loc[
    (df_concatenado_clientes['id_cliente'] == 47)
    & (df_concatenado_clientes['nome'] == 'Carlos Neto'),
    'id_cliente'] = 401

print(df_concatenado_clientes['id_cliente'].value_counts())
print(df_concatenado_clientes[df_concatenado_clientes['id_cliente'].isin([47, 401])])

id_cliente
1      1
2      1
3      1
4      1
5      1
      ..
316    1
317    1
318    1
319    1
320    1
Name: count, Length: 321, dtype: int64
     id_cliente           nome cidade_origem estado_origem faixa_etaria  \
46           47  Wesley Araujo        Recife            PE          65+   
300         401    Carlos Neto       Niterói            RJ        36-45   

      tipo_cliente  
46   PESSOA FÍSICA  
300    CORPORATIVO  


In [134]:
df_atual = df_concatenado_clientes

relatorio_final(df_concatenado_clientes_original, df_atual)

      RELATÓRIO FINAL DE AUDITORIA
Registros iniciais:        321
Registros finais:          321
Registros removidos:         0
Colunas originais:           6
Colunas finais:              6
Nulos restantes:            65
Duplicatas restantes:        0


### 2.3 base: Canais_venda e Unidades

Base: Canais_Venda: padronização da grafia do nome do canal
Ota - Online Travel Agency

In [135]:
#relatorio_inicial(canais_original)
canais['nome_canal'] = canais['nome_canal'].str.strip().str.upper()

Base: Unidades: 
1) Padronização das grafias nas colunas região e categoria do hotel
2) Isere, através da função MAP, o número de quartos faltantes nas colunas "total de quartos"

In [136]:
relatorio_inicial(df_concatenado_unidades_original)

df_concatenado_unidades['regiao'] = (
    df_concatenado_unidades['regiao']
    .replace('CAP', 'Capital')
    .str.strip()                
    .str.title()                
)

df_concatenado_unidades['categoria_hotel'] = (
    df_concatenado_unidades['categoria_hotel']
    .str.lower()                           
    .str.replace('cinco', '5')             
    .str.replace('três', '3')           
    .str.replace('*', ' estrelas')
    .str.title()
)

df_concatenado_unidades.tail(15)

ETAPA 1 - AUDITORIA

[1] COLUNAS DA BASE:
['id_unidade', 'nome_unidade', 'cidade', 'regiao', 'categoria_hotel', 'num_quartos_total']

[2] INFO DA BASE:
<class 'pandas.DataFrame'>
RangeIndex: 13 entries, 0 to 12
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id_unidade         13 non-null     int64  
 1   nome_unidade       13 non-null     str    
 2   cidade             13 non-null     str    
 3   regiao             13 non-null     str    
 4   categoria_hotel    13 non-null     str    
 5   num_quartos_total  11 non-null     float64
dtypes: float64(1), int64(1), str(4)
memory usage: 756.0 bytes

[3] TIPOS DE DADOS (dtypes):
id_unidade             int64
nome_unidade             str
cidade                   str
regiao                   str
categoria_hotel          str
num_quartos_total    float64
dtype: object

[4] ESTATÍSTICAS DESCRITIVAS (Numéricas):
       id_unidade  num_quartos_total
count    13.

,id_unidade,nome_unidade,cidade,regiao,categoria_hotel,num_quartos_total
0,1,NaraHoteis Ipanema,Rio de Janeiro,Capital,5 Estrelas,80.0
1,2,NaraHoteis Barra da Tijuca,Rio de Janeiro,Capital,4 Estrelas,120.0
2,3,NaraHoteis Centro,Rio de Janeiro,Capital,3 Estrelas,60.0
3,4,NaraHoteis Santa Teresa,Rio de Janeiro,Capital,4 Estrelas,45.0
4,5,NaraHoteis Nova Iguaçu Centro,Nova Iguaçu,Baixada Fluminense,3 Estrelas,50.0
5,6,NaraHoteis Nova Iguaçu Park,Nova Iguaçu,Baixada Fluminense,3 Estrelas,55.0
6,7,NaraHoteis Petrópolis,Petrópolis,Serra,5 Estrelas,NaN
7,8,NaraHoteis Teresópolis,Teresópolis,Serra,4 Estrelas,60.0
8,9,NaraHoteis Friburgo,Nova Friburgo,Serra,4 Estrelas,55.0
9,10,NaraHoteis Paraty,Paraty,Costa Verde,5 Estrelas,NaN


In [137]:
quartos_faltantes = {
    7: 70.0,
    10: 75.0
}

df_concatenado_unidades['num_quartos_total'] = (
    df_concatenado_unidades['num_quartos_total']
    .fillna(df_concatenado_unidades['id_unidade'].map(quartos_faltantes))
)

print(df_concatenado_unidades['num_quartos_total'].isnull().sum())

df_concatenado_unidades

0


,id_unidade,nome_unidade,cidade,regiao,categoria_hotel,num_quartos_total
0,1,NaraHoteis Ipanema,Rio de Janeiro,Capital,5 Estrelas,80.0
1,2,NaraHoteis Barra da Tijuca,Rio de Janeiro,Capital,4 Estrelas,120.0
2,3,NaraHoteis Centro,Rio de Janeiro,Capital,3 Estrelas,60.0
3,4,NaraHoteis Santa Teresa,Rio de Janeiro,Capital,4 Estrelas,45.0
4,5,NaraHoteis Nova Iguaçu Centro,Nova Iguaçu,Baixada Fluminense,3 Estrelas,50.0
5,6,NaraHoteis Nova Iguaçu Park,Nova Iguaçu,Baixada Fluminense,3 Estrelas,55.0
6,7,NaraHoteis Petrópolis,Petrópolis,Serra,5 Estrelas,70.0
7,8,NaraHoteis Teresópolis,Teresópolis,Serra,4 Estrelas,60.0
8,9,NaraHoteis Friburgo,Nova Friburgo,Serra,4 Estrelas,55.0
9,10,NaraHoteis Paraty,Paraty,Costa Verde,5 Estrelas,75.0


* Aplicando a função para Relatório Final

In [138]:
df_original = df_concatenado_unidades_original
df_atual = df_concatenado_unidades

relatorio_final(df_original, df_atual)

      RELATÓRIO FINAL DE AUDITORIA
Registros iniciais:         13
Registros finais:           13
Registros removidos:         0
Colunas originais:           6
Colunas finais:              6
Nulos restantes:             0
Duplicatas restantes:        0


Exportando CSV

In [140]:
canais.to_csv('canais_tratada.csv', index=False)
df_concatenado_clientes.to_csv('clientes_tratada.csv', index=False)
df_concatenado_quartos.to_csv('quartos_tratada.csv', index=False)
df_concatenado_funcionarios.to_csv('funcionarios_tratada.csv', index=False)
df_concatenado_reservas.to_csv('reservas_tratada.csv', index=False)
df_concatenado_unidades.to_csv('unidades_tratada.csv', index=False)
